lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -816,3 +829,243 @@ def get_provider_vault_token(
                     label=self.AUTOUNSEAL_TOKEN_SECRET_LABEL,
                 )
         return external_vault.token
+
+
+class PKIManager:
+    """Encapsulates the business logic for managing PKI certificates in Vault from a Charm."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        vault_client: VaultClient,
+        common_name: str,
+        mount_point: str,
+        role_name: str,
+        vault_pki: TLSCertificatesProvidesV4,
+        tls_certificates_pki: TLSCertificatesRequiresV4,
+    ):
+        """Create a new PKIManager object.
+
+        Args:
+            charm: The charm this manager is associated with
+            vault_client: The Vault client object
+            common_name: The common name for the certificates allowed by the PKI backend
+            mount_point: The mount point in Vault for the PKI backend
+            role_name: The role name for the PKI backend
+            vault_pki: The vault_pki provider relation helper library
+            tls_certificates_pki: The tls_certificates_pki requirer relation helper library
+        """
+        self.validate_input(common_name=common_name)
+        self._vault_client = vault_client
+        self._juju_facade = JujuFacade(charm)
+        self._common_name = common_name
```

Comment:
Should we have the CertificateRequestAttributes be the parameter instead of the common_name? This will allow requesting certs with more fields than just the common name without changing anything here.

Reply:
That probably makes sense. let me take a peek.

Reply:
Definitely makes more sense, since the manager it isn't the thing putting the request in the databag in the first place. I'll make this change.

Reply:
https://github.com/canonical/vault-k8s-operator/pull/559/commits/04162cb3c26d73c578a49092153cba2a3e30c5fd